crates/net/eth-wire/src/handshake.rs

```diff
@@ -0,0 +1,193 @@
+use crate::{
+    errors::{EthHandshakeError, EthStreamError, P2PStreamError},
+    ethstream::MAX_STATUS_SIZE,
+    CanDisconnect,
+};
+use bytes::{Bytes, BytesMut};
+use derive_more::with_trait::Debug;
+use futures::{Sink, SinkExt, Stream};
+use reth_eth_wire_types::{
+    DisconnectReason, EthMessage, EthNetworkPrimitives, ProtocolMessage, Status,
+};
+use reth_ethereum_forks::ForkFilter;
+use reth_primitives_traits::GotExpected;
+use std::{future::Future, pin::Pin, time::Duration};
+use tokio::time::timeout;
+use tokio_stream::StreamExt;
+use tracing::{debug, trace};
+
+/// A trait that knows how to perform the P2P handshake.
+pub trait EthRlpxHandshake: Debug + Send + Sync + 'static {
+    /// Perform the P2P handshake.
+    fn handshake<'a>(
+        &'a self,
+        unauth: &'a mut dyn UnauthEth,
+        status: Status,
+        fork_filter: ForkFilter,
+        timeout_limit: Duration,
+    ) -> Pin<Box<dyn Future<Output = Result<Status, EthStreamError>> + 'a + Send>>;
+}
+
+/// An unauthenticated stream that can send and receive messages.
+pub trait UnauthEth:
+    Stream<Item = Result<BytesMut, P2PStreamError>>
+    + Sink<Bytes, Error = P2PStreamError>
+    + CanDisconnect<Bytes>
+    + Unpin
+    + Send
+{
+}
+
+impl<T> UnauthEth for T where
+    T: Stream<Item = Result<BytesMut, P2PStreamError>>
+        + Sink<Bytes, Error = P2PStreamError>
+        + CanDisconnect<Bytes>
+        + Unpin
+        + Send
+{
+}
+
+/// The Ethereum P2P handshake.
+#[derive(Debug, Default, Clone)]
+pub struct EthHandshake;
+
+impl EthRlpxHandshake for EthHandshake {
+    fn handshake<'a>(
+        &'a self,
+        unauth: &'a mut dyn UnauthEth,
+        status: Status,
+        fork_filter: ForkFilter,
+        timeout_limit: Duration,
+    ) -> Pin<Box<dyn Future<Output = Result<Status, EthStreamError>> + 'a + Send>> {
+        Box::pin(async move {
+            timeout(timeout_limit, eth_handshake(unauth, status, fork_filter))
+                .await
+                .map_err(|_| EthStreamError::StreamTimeout)?
+        })
+    }
+}
+
+/// Performs the `eth` rlpx protocol handshake using the given input stream.
+pub async fn eth_handshake(
+    unauth: &mut dyn UnauthEth,
+    status: Status,
+    fork_filter: ForkFilter,
+) -> Result<Status, EthStreamError> {
```

Comment:
we now have this twice, we should remove this duplication from the `UnauthedEthStream::handshake` function